Preserve order of decls through pickle/unpickle

retronym opened the pull request:
The pickle format does not explicitly enccode the order of decls. Instead,
symbols are entered into an index in the order that they are found by the
pickler, either as a definition or as a reference. During unpickling,
symbols are read and entered into the owner's decls in that order.

This is a cause of unstable compiler output: a class that mixes in members
from some trait will have a different order if it is compiled jointly with
/ separately from that trait.

This commit modifies the pickler with an initial pass that reserves index
entries for all declarations in the declaration order.

The pickle format and the unpickler are unchanged.

References scala/scala-dev#405

retronym committed:
Refactor pickle phase

Small refactoring designed to make the subsequent
commit more reviewable.

Tightens up the definition of "companion"-s so that
we no longer add a type alias in a package object into
the pickle of a same-named module.

retronym committed:
scalap should not print class type param annots in decls

The annotation removed in this diff was actually from `R`!

adriaanm commented on src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala:
```diff
@@ -35,24 +35,50 @@ abstract class Pickler extends SubComponent {
     def apply(unit: CompilationUnit) {
       def pickle(tree: Tree) {
-        def add(sym: Symbol, pickle: Pickle) = {
+        def add(sym: Symbol, pickle: Pickle) = withPickle(sym, pickle){ pickle =>
```
`withPickle` is called only once -- should we just leave it inlined? The `shouldPickle` condition does occur twice, though.

adriaanm replied:
ah, maybe the plan was to factor out the
```
pickle.putSymbol(sym)
currentRun.symData(sym) = pickle
```

part?

retronym replied:
Oops, `def add` itself was dead code from a failed refactoring attempt. I'll push a cleanup.

adriaanm commented on src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala:
```diff
@@ -35,24 +35,50 @@ abstract class Pickler extends SubComponent {
+            val shouldPickleSym = shouldPickle(sym)
+            if (shouldPickleSym) {
+              val pickle = new Pickle(sym)
+              def reserveDeclEntries(sym: Symbol): Unit = {
```
How about:
```
def reserveDeclEntries(sym: Symbol): Unit =
  if (sym.isClass) sym.info.decls.foreach(reserveDeclEntries)
  else if (sym.isModule) reserveDeclEntries(sym.moduleClass)
  else pickle.reserveEntry(sym)
```
?

retronym replied:
Actually, we need to reserve the slots for the class/module symbols themselves, prior to recursing into their decls. Pushed.

adriaanm reviewed (changes requested):
A few suggestions / questions.

retronym committed:
Preserve order of decls through pickle/unpickle

The pickle format does not explicitly enccode the order of
decls. Instead, symbols are entered into an index in the
order that they are found by the pickler, either as a definition
or as a reference. During unpickling, symbols are read and entered
into the owner's decls in that order.

This is a cause of unstable compiler output: a class that mixes
in members from some trait will have a different order if it is
compiled jointly with / separately from that trait.

This commit modifies the pickler with an initial pass that reserves
index entries for all declarations in the declaration order.

The pickle format and the unpickler are unchanged.

retronym commented on src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala:
```diff
@@ -35,24 +35,28 @@ abstract class Pickler extends SubComponent {
       def pickle(tree: Tree) {
-        def add(sym: Symbol, pickle: Pickle) = {
-          if (currentRun.compiles(sym) && !currentRun.symData.contains(sym)) {
-            debuglog("pickling " + sym)
```
I didn't think this logging was worth keeping.

retronym commented:
/synch

retronym commented:
/rebuild

retronym commented:
/synch

retronym commented:
/rebuild

retronym commented:
/synch

SethTisue commented:
/synch

retronym commented:
rebased
